src/main/java/org/commcare/suite/model/StackFrameStep.java

```diff
@@ -110,6 +110,10 @@ public ExternalDataInstance getXmlInstance() {
         return xmlInstance;
     }
 
+    public void setXmlInstance(ExternalDataInstance xmlInstance) {
```

Comment:
I don't really like making this mutable but could not come up with a better way either.

Reply:
If mutability is the only practical way to get the workflow right, one option to make it safer is to control how mutable the data is in the lifecycle of the frame step. IE: If the instance should only be set once and in a certain order, after it is set you could "seal" the object so that future attempts to change it through the same interface fail loudly. 

That would let you continue to guarantee certain aspects of how the classes work. For most purposes it's still effectively immutable if when you reference it it's either a consistent value or it's uninitialized and errors.